src/Contacts/PhoneNumber.m

```diff
@@ -115,6 +119,75 @@ + (PhoneNumber *)tryParsePhoneNumberFromUserSpecifiedText:(NSString *)text {
     return [self phoneNumberFromUserSpecifiedText:sanitizedString];
 }
 
++ (NSArray *)tryParsePhoneNumbersFromsUserSpecifiedText:(NSString *)text
+                                      clientPhoneNumber:(NSString *)clientPhoneNumber {
+    assert(text != nil);
+    
+    text = [text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
+    if ([text isEqualToString:@""]) {
+        return nil;
+    }
+    
+    NSString *sanitizedString = [self removeFormattingCharacters:text];
+    assert(sanitizedString != nil);
+    
+    NSMutableArray *result = [NSMutableArray new];
+    NSMutableSet *phoneNumberSet = [NSMutableSet new];
+    void (^tryParsingWithCountryCode)(NSString *, NSString *) = ^(NSString *text,
+                                                      NSString *countryCode) {
+        PhoneNumber *phoneNumber = [PhoneNumber phoneNumberFromText:text
+                                                          andRegion:countryCode];
+        if (phoneNumber &&
+            ![phoneNumberSet containsObject:[phoneNumber toE164]]) {
+            [result addObject:phoneNumber];
+            [phoneNumberSet addObject:[phoneNumber toE164]];
+        }
+    };
+
+    if ([sanitizedString hasPrefix:@"+"]) {
+        // If the text starts with "+", just parse it.
+        tryParsingWithCountryCode(sanitizedString, [self defaultRegionCode]);
+
+        return result;
+    }
+
+    // Order matters; better results should appear first so
+    // prefer matches using this client's phone number.
+    if (clientPhoneNumber.length > 0) {
```

Comment:
Well, I was trying to leave this open to be used when the app is non-registered but I guess I'll add an assert for now.